driver/src/test/scala/za/co/absa/hyperdrive/driver/drivers/KafkaToKafkaDeduplicationAfterRetryDockerTest.scala

```diff
@@ -0,0 +1,246 @@
+ * to identify messages across the topics (See transformer.[kafka.deduplicator].source.id.columns
+ * and transformer.[kafka.deduplicator].destination.id.columns)
+ *
+ * Finally, the destination topic is expected to contain all messages from the source topic
+ * exactly once. Without the deduplication transformer, the 41st-50th messages would be duplicated.
+ */
+class KafkaToKafkaDeduplicationAfterRetryDockerTest extends FlatSpec with Matchers with SparkTestBase with BeforeAndAfter {
+  import scala.collection.JavaConverters._
+
+  private val fs = FileSystem.get(spark.sparkContext.hadoopConfiguration)
+  private val baseDir = TempDirectory("hyperdriveE2eTest").deleteOnExit()
+  private val baseDirPath = baseDir.path.toUri
+  private val checkpointDir = s"$baseDirPath/checkpoint"
+
+  behavior of "CommandLineIngestionDriver"
+
+  private def schemaV1String(name: String) =
+    raw"""{"type": "record", "name": "$name", "fields": [
+         |{"type": "int", "name": "record_id"},
+         |{"type": "string", "name": "value_field", "nullable": false}
+         |]}""".stripMargin
+
+  private def schemaV2String(name: String) =
+    raw"""{"type": "record", "name": "$name", "fields": [
+         |{"type": "int", "name": "record_id"},
+         |{"type": ["null", "string"], "name": "value_field", "nullable": true}
+         |]}""".stripMargin
+
+  private def sendData(producer: KafkaProducer[GenericRecord, GenericRecord], records: Seq[GenericRecord], topic: String, partitions: Int): Unit = {
+    records.zipWithIndex.foreach {
+      case (record, i) =>
+        val partition = i % partitions
+        val producerRecord = new ProducerRecord[GenericRecord, GenericRecord](topic, partition, null, record)
+        producer.send(producerRecord)
+    }
+  }
+
+  private def createTopic(kafkaSchemaRegistryWrapper: KafkaSchemaRegistryWrapper, topicName: String, partitions: Int): Unit = {
+    val config = new Properties()
+    config.put(AdminClientConfig.BOOTSTRAP_SERVERS_CONFIG, kafkaSchemaRegistryWrapper.kafka.getBootstrapServers)
+    val localKafkaAdmin = AdminClient.create(config)
+    val replication = 1.toShort
+    val topic = new NewTopic(topicName, partitions, replication)
+    val topicCreationFut = localKafkaAdmin.createTopics(util.Arrays.asList(topic)).all()
+    while(!topicCreationFut.isDone) {}
+  }
+
+  it should "execute the whole kafka-to-kafka pipeline" in {
+    // given
+    val kafkaSchemaRegistryWrapper = new KafkaSchemaRegistryWrapper
+    val sourceTopic = "deduplication_src"
+    val destinationTopic = "deduplication_dest"
+    val sourceTopicPartitions = 5
+    val destinationTopicPartitions = 3
+    val schemaManager = SchemaManagerFactory.create(Map("schema.registry.url" -> kafkaSchemaRegistryWrapper.schemaRegistryUrl))
+    val subject = SchemaSubject.usingTopicNameStrategy(sourceTopic)
+    val parserV1 = new Parser()
+    val schemaV1 = parserV1.parse(schemaV1String(sourceTopic))
+    val parserV2 = new Parser()
```

Comment:
Just a silly comment, but do you need two parsers?

Reply:
Of course, this is not a silly question. If I reuse the same parser, I get 
```
org.apache.avro.SchemaParseException: Can't redefine: deduplication_src
	at org.apache.avro.Schema$Names.put(Schema.java:1128)
	at org.apache.avro.Schema$Names.add(Schema.java:1123)
	at org.apache.avro.Schema.parse(Schema.java:1253)
	at org.apache.avro.Schema$Parser.parse(Schema.java:1032)
```
So this parser is stateful and allows only one schema per name, that's why I need a separate parser for schema version 2

Reply:
Of course, it is a mutable class.